#!/usr/bin/env Rscript

cat(paste("\n[",Sys.time(),"]\tConvert input target regions from BED format to annotated VCF-like format (one targeted genomic position per each row)","\n"))

library(parallel)
library(data.table)
options(scipen=99999)

args <- commandArgs(trailingOnly = TRUE)
if(length(args)!=5){
  message("\nERROR:\t5 arguments required")
  message("\nUSAGE:\tRscript PrepareTargetPositions.R <bed> <fasta> <dbsnp.bychrom> <outdir> <threads>\n")
  cat(paste("\t[1] bed\t\t\tTarget regions in BED format",
            "\t[2] fasta\t\tReference Genome in fasta format used to align BAM files",
            "\t[3] dbsnp.bychrom\tFolder 00-All.vcf.gz.bychrom generated by utility/dbSNPPER.R",
            "\t[4] outdir\t\tOutput folder in which subfolder TargetPositions will be created",
            "\t[5] threads\t\tNumber of threads",sep="\n"),"\n\n")  
  quit()
}

targetbed = args[1]
ReferenceFasta = args[2]  
dbsnpfolder = args[3]
outdir = args[4]
Ncores = args[5]

timestart <- proc.time()

if(!file.exists(file.path(outdir, "TargetPositions"))){
  dir.create(file.path(outdir, "TargetPositions"), showWarnings = TRUE)
}
setwd(file.path(outdir, "TargetPositions"))

fromListToDF <- function(inputList){
  if (is.null(inputList)){return(NULL)}
  #check if some is null and remove
  nullPositions <- which(sapply(inputList,is.null))
  if (length(nullPositions) > 0){
    inputList <- inputList[-nullPositions]    
  }
  firstEl <- inputList[[1]]  
  inputList <- lapply(inputList, function(x){ matrix(unlist(x), ncol=ncol(x))} )
  outDF <- as.data.frame(do.call(rbind, inputList),stringsAsFactors=F)
  colnames(outDF) <-names(firstEl)
  
  for(idx in c(1:ncol(outDF))){
    if (class(firstEl[[idx]]) == "logical"){       
      if (is.na(firstEl[[idx]])){
        class(outDF[[idx]]) <- "numeric"
      }else if (outDF[1,idx] == "TRUE" ||outDF[1,idx] == "FALSE"  ){
        outDF[,idx] <- as.logical(outDF[,idx]) * 1      
      }
      class(outDF[[idx]]) <- "numeric"
    }else{
      class(outDF[[idx]]) <- class(firstEl[[idx]])
    }
  }
  return(outDF)
}

GCperc <- function(fastaseq){
  fastaseq = toupper(fastaseq)
  gc = round(sum(fastaseq=="G"|fastaseq=="C")/length(fastaseq)*100,2)
  return(gc)  
}

mappability <- function(id,bed,kmer){
  this = bed[id,,drop=F]
  kmer = as.character(kmer)
  chromosome = gsub(this[,1],pattern = "chr",replacement = "")
  chr = paste0("-chrom=chr",chromosome)
  start = paste0("-start=",this[,2])
  end = paste0("-end=",this[,3])
  bigWigToBedGraph = "/elaborazioni/sharedCO/ISMB_ECCB_2015/Data/Mappability/BigWig/bigWigToBedGraph"
  encodedata = list.files("/elaborazioni/sharedCO/ISMB_ECCB_2015/Data/Mappability/BigWig",pattern = paste0(kmer,"mer.bigWig"),full.names = T)
  out = paste0("/elaborazioni/sharedCO/ISMB_ECCB_2015/Data/Mappability/BigWig/tmpdata/",paste0(this[1:3],collapse = "_"),".bedGraph")
  cmdmap =  paste(bigWigToBedGraph,encodedata,chr,start,end,out)
  system(cmdmap)
  mapp = read.delim(out,as.is=T,header=F)
  this$mean_mappability <- round(weighted.mean(mapp$V4,w = mapp$V3-mapp$V2),5)
  clean = paste("rm",out)
  system(clean)
  return(this)
}

uniqueness <- function(id,bed,kmer){
  this = bed[id,,drop=F]
  kmer = as.character(kmer)
  chromosome = gsub(this[,1],pattern = "chr",replacement = "")
  chr = paste0("-chrom=chr",chromosome)
  start = paste0("-start=",this[,2])
  end = paste0("-end=",this[,3])
  bigWigToBedGraph = "/elaborazioni/sharedCO/ISMB_ECCB_2015/Data/Mappability/BigWig/bigWigToBedGraph"
  encodedata = list.files("/elaborazioni/sharedCO/ISMB_ECCB_2015/Data/Mappability/BigWig",pattern = paste0(kmer,"bp.bigWig"),full.names = T)
  out = paste0("/elaborazioni/sharedCO/ISMB_ECCB_2015/Data/Mappability/BigWig/tmpdata/",paste0(this[1:3],collapse = "_"),".bedGraph")
  cmduniq =  paste(bigWigToBedGraph,encodedata,chr,start,end,out)
  system(cmduniq)
  uniq = read.delim(out,as.is=T,header=F)
  this$mean_uniq <- round(weighted.mean(uniq$V4,w = uniq$V3-uniq$V2),5)
  clean = paste("rm",out)
  system(clean)
  return(this)
}

getpos <- function(id,bed,fasta){
  chrom = bed[id,1] 
  start = bed[id,2]
  end = bed[id,3]
  pos = seq(start+1,end,1)
  sequence = unlist(strsplit(fasta[which(fasta$chr==chrom & fasta$start==start & fasta$end==end),4],split = "*"))
  if(length(pos)!=length(sequence)){
    cat(paste("ERROR! Fasta sequence length different from target BED region:\t",id))
    quit()
  }
  gc = GCperc(fastaseq = sequence)
  map = mappability(id = id,bed = bed,kmer = 24)
  uniq = uniqueness(id = id,bed = bed,kmer = 20)
  # merge info
  regionpos = data.frame(chr=chrom,pos=pos,ref=toupper(sequence),gc=gc,map=map$mean_mappability,uniq=uniq$mean_uniq,stringsAsFactors = F)
  return(regionpos)
}

cat(paste("[",Sys.time(),"]\tLoading targets in :",targetbed,"\n"))
bed <- read.delim(targetbed,as.is=T,comment.char = "#",header = F)
bed <- unique(bed)
dupstarget = which(duplicated(bed[,1:3]))
if(length(dupstarget)>0){
  bed = bed[-dupstarget,]
}
cat(paste("[",Sys.time(),"]\tn. of target regions found :",nrow(bed),"\n"))
cat(paste("[",Sys.time(),"]\tListing dbSNP files in folder :",dbsnpfolder,"\n"))
dbsnp_chroms = list.files(dbsnpfolder,full.names = T)

# get fasta sequence for each target region 
fileout = basename(gsub(targetbed,pattern = ".bed",replacement = "_tmp.bed")) 
target <- bed 
write.table(target,file = fileout,quote = F,col.names = F,row.names = F,sep="\t")
# sort bed file [0-based]
fileoutsort = gsub(fileout,pattern = ".bed",replacement = ".sort.bed")
cmd = paste("bedtools sort -i",fileout,">",fileoutsort)
system(cmd)
system(paste("rm",fileout))
fastabed = basename(gsub(fileoutsort,pattern = ".bed",replacement = ".fasta"))
cmd = paste("bedtools getfasta -fi",ReferenceFasta,"-bed",fileoutsort,"-fo",fastabed,"-tab")
system(cmd)
fastatab = gsub(fastabed,pattern = ".fasta",replacement = ".tab.fasta")
cmd = paste("tr ':|-' \\\t <",fastabed,">",fastatab)
system(cmd)
system(paste("rm",fastabed))
fasta = read.delim(fastatab,header = F,as.is=T)
fasta = unique(fasta)
colnames(fasta) <- c("chr","start","end","sequence")
system(paste("rm",fileoutsort))
system(paste("mv",fastatab,"TargetPositions.fasta"))

# Split by chromosomes
CHR = sort(unique(bed[,1]))
totpos = data.frame(chromosome = CHR,bp_covered = 0,row.names = CHR,stringsAsFactors = F)
for(chrom in CHR){
  cat(paste("[",Sys.time(),"]\tchromosome:",chrom,"\n"))
  chrombed = bed[which(bed[,1]==chrom),]
  # import dbsnp chromosome file
  PATTERN = paste0("_chr",gsub(chrom,pattern = "chr",replacement = ""),".intersectbed.snv.single.vcf")
  dbsnpfile = grep(dbsnp_chroms,pattern = PATTERN,value = T)
  chromdbsnp <- fread(dbsnpfile,header = F,sep = "\t",select = c(1:5))
  setnames(chromdbsnp,c("chr","pos","dbsnp","ref","alt"))
  if(grepl("chr",chrom)){
    chromdbsnp$chr <- paste0("chr",chromdbsnp$chr)
  }
  # BED to VCF
  vcf <- fromListToDF(mclapply(seq(1,nrow(chrombed),1),getpos,bed=chrombed,fasta=fasta,mc.cores = Ncores))
  vcf <- unique(as.data.table(vcf))
  chromdbsnp$chr = as.character(chromdbsnp$chr)
  vcf$chr = as.character(vcf$chr)
  TargetPositions = merge(x = vcf,y = chromdbsnp,by=c("chr","pos","ref"),all.x = T)
  TargetPositions = TargetPositions[,c("chr","pos","ref","dbsnp","gc","map","uniq"),with=FALSE]
  # merge rs IDs associated to the same locus
  TargetPositions = as.data.frame(TargetPositions)
  dupsids = which(duplicated(TargetPositions$pos))
  cat(paste("[",Sys.time(),"]\tn. of loci with multiple rs IDs :",length(dupsids),"\n"))
  if(length(dupsids)>0){
    TargetPositions = TargetPositions[-dupsids,]
  }
  TargetPositions = TargetPositions[with(TargetPositions,order(pos)),]
  totpos[as.character(chrom),"bp_covered"] <- nrow(TargetPositions)
  filename = paste0("TargetPositions_chr",gsub(chrom,pattern = "chr",replacement = ""),".RData")
  assign(x = "chromTargetPositions",TargetPositions)
  cat(paste("[",Sys.time(),"]\tSaving data for this chromosome","\n"))
  save(list ="chromTargetPositions",file = filename)
}
cat(paste("[",Sys.time(),"]\tWriting data table : bpcovered.tsv","\n"))
bpcovtotal = sum(totpos$bp_covered,na.rm = T)
tot = data.frame(chromosome="TOTAL",bp_covered=bpcovtotal,stringsAsFactors = F)
totpos = rbind(totpos,tot)
totpos$bp_covered_fraction = totpos$bp_covered/bpcovtotal
totpos$chromosome = gsub(totpos$chromosome,pattern = "chr",replacement = "")
write.table(totpos,file = "bpcovered.tsv",quote = F,sep="\t",col.names = T,row.names = F)

proc.time()-timestart
